packages/playground/blueprints/src/lib/zip.ts

```diff
@@ -0,0 +1,753 @@
+	) => boolean = DEFAULT_PREDICATE
+) {
+	if (predicate === DEFAULT_PREDICATE) {
+		// If we're not filtering the zip contents, let's just
+		// grab the entire zip.
+		const response = await fetch(url);
+		return iterateZipStream(response.body!);
+	}
+
+	const contentLength = await getContentLength(url);
+	if (contentLength <= PREFER_RANGES_IF_FILE_LARGER_THAN) {
+		// If the zip is small enough, let's just grab it.
+		const response = await fetch(url);
+		return iterateZipStream(response.body!);
+	}
+
+	// Ensure ranges query support:
+	// Fetch one byte
+	const response = await fetch(url, {
+		headers: {
+			Range: 'bytes=0-0',
+			'Accept-Encoding': 'none',
+		},
+	});
+
+	// Fork the stream so that we can reuse it in case
+	// bytes are unsupported and we're now streaming the
+	// entire file
+	const [peekStream, responseStream] = response.body!.tee();
+
+	// Read from the forked stream and confirm we only
+	// got a single byte.
+	const peekReader = peekStream.getReader();
+	const { value: peekBytes } = await peekReader.read();
+	const { done: peekDone } = await peekReader.read();
+	// Close the forked stream
+	peekReader.releaseLock();
+	peekStream.cancel();
+
+	// Confirm our Range query worked as intended:
+	const rangesSupported = peekBytes?.length === 1 && peekDone;
+	if (!rangesSupported) {
+		// Uh-oh, we're actually streaming the entire file.
+		// Let's reuse the forked stream as our response stream.
+		return iterateZipStream(responseStream);
+	}
+
+	// We're good, let's clean up the other branch of the response stream.
+	responseStream.cancel();
+	const source = await createFetchSource(url, contentLength);
+	return centralDirectoryEntries(source)
+		.pipeThrough(filterStream(predicate))
+		.pipeThrough(partitionNearbyEntries())
+		.pipeThrough(
+			fetchPartitionedEntries(source)
+		) as IterableReadableStream<FileEntry>;
+}
+
+export function iterateZipStream(
+	stream: ReadableStream<Uint8Array>,
+	predicate: (
+		dirEntry: CentralDirectoryEntry | FileEntry
+	) => boolean = DEFAULT_PREDICATE
+) {
+	return zipEntries(stream)
+		.pipeThrough(
+			filterStream(({ signature }) => signature === SIGNATURE_FILE)
+		)
+		.pipeThrough(
+			filterStream(predicate as any)
+		) as IterableReadableStream<FileEntry>;
+}
+
+export function zipEntries(stream: ReadableStream<Uint8Array>) {
+	return new ReadableStream<ZipEntry>({
+		async pull(controller) {
+			try {
+				const entry = await readNextEntry(stream);
+				if (!entry) {
+					controller.close();
+					return;
+				}
+				controller.enqueue(entry);
+			} catch (e) {
+				console.error(e);
+				controller.error(e);
+				throw e;
+			}
+		},
+	}) as IterableReadableStream<ZipEntry>;
+}
+
+async function getContentLength(url: string) {
+	return await fetch(url, { method: 'HEAD' })
+		.then((response) => response.headers.get('Content-Length'))
+		.then((contentLength) => {
+			if (!contentLength) {
+				throw new Error('Content-Length header is missing');
```

Comment:
perfect! I came here to verify that we account for the content-length not being here.

might consider negative content lengths and invalid parses, as some servers might send weird lengths or invalid lengths

Reply:
Good note! I just added that in https://github.com/WordPress/wordpress-playground/pull/851/commits/46c36ebd364119c34dabea1b1a08bca770a00e8d:

```ts

			const parsedLength = parseInt(contentLength, 10);
			if (isNaN(parsedLength) || parsedLength < 0) {
				throw new Error('Content-Length header is invalid');
			}
			return parsedLength;
```